subprojects/plugin-development/src/integTest/groovy/org/gradle/plugin/devel/tasks/AbstractPluginValidationIntegrationSpec.groovy

```diff
@@ -0,0 +1,766 @@
+            import java.io.File;
+
+            public class MyTask extends DefaultTask {
+                @Input
+                public String notAGetter() {
+                    return "not-a-getter";
+                }
+
+                @Nested
+                public Options getOptions() {
+                    return new Options();
+                }
+
+                public static class Options {
+                    @Input
+                    public String notANestedGetter() {
+                        return "not-a-nested-getter";
+                    }
+                }
+
+                @TaskAction
+                public void doStuff() { }
+            }
+        """
+
+        expect:
+        assertValidationFailsWith(
+            "Type 'MyTask\$Options': non-property method 'notANestedGetter()' should not be annotated with: @Input": WARNING,
+            "Type 'MyTask': non-property method 'notAGetter()' should not be annotated with: @Input": WARNING,
+        )
+    }
+
+    def "detects annotations on setter methods"() {
+        javaTaskSource << """
+            import org.gradle.api.*;
+            import org.gradle.api.tasks.*;
+            import java.io.File;
+
+            public class MyTask extends DefaultTask {
+                @Input
+                public void setWriteOnly(String value) {
+                }
+
+                public String getReadWrite() {
+                    return "read-write property";
+                }
+
+                @Input
+                public void setReadWrite(String value) {
+                }
+
+                @Nested
+                public Options getOptions() {
+                    return new Options();
+                }
+
+                public static class Options {
+                    @Input
+                    public void setWriteOnly(String value) {
+                    }
+
+                    @Input
+                    public String getReadWrite() {
+                        return "read-write property";
+                    }
+
+                    @Input
+                    public void setReadWrite(String value) {
+                    }
+                }
+
+                @TaskAction
+                public void doStuff() { }
+            }
+        """
+
+        expect:
+        assertValidationFailsWith(
+            "Type 'MyTask\$Options': setter method 'setReadWrite()' should not be annotated with: @Input": WARNING,
+            "Type 'MyTask\$Options': setter method 'setWriteOnly()' should not be annotated with: @Input": WARNING,
+            "Type 'MyTask': property 'readWrite' is not annotated with an input or output annotation.": WARNING,
+            "Type 'MyTask': setter method 'setReadWrite()' should not be annotated with: @Input": WARNING,
+            "Type 'MyTask': setter method 'setWriteOnly()' should not be annotated with: @Input": WARNING,
+        )
+    }
+
+    @Requires(TestPrecondition.JDK8_OR_LATER)
```

Comment:
Why do we need this requires? Gradle doesn't run on JDK < 8.

It also seems that this task only belongs to the `ValidatePluginsIntegrationTest`.

Reply:
I removed the `@Requires` but why do you think the test should stay in `ValidatePluginsIntegrationTest`?

Reply:
The tests should only be in `ValidatePluginsIntegrationTest`, since classloading issues for validated user types can only occur for the static analysis. For the runtime analysis, we have the actual objects, so the classes were loaded already.

Reply:
Fair point. I moved them back.